############################################################
# This script is used for metagenomics downstream analysis #
############################################################

setwd("D:/Project/Microbiome_Influenza/")
library(dplyr)
library(reshape2)
library(ggplot2)
library(ggpubr)
library(vegan)
library(ape)

###################################################################
## Abundance of dominant bacteria at time point 1 at phylum rank ##
## The input file is generated by MEGAN version 6
phylum                  <- read.csv("Megan_phylum.csv", header = T, stringsAsFactors = FALSE)
colnames(phylum)        <- c("phylum",paste0("T1_",1:10),paste0("T2_",1:10), paste0("T3_",1:10))

phylum.T1               <- select(phylum,c(1,starts_with("T1_")))
phylum.T1$v1_sum        <- rowSums(phylum.T1[,-1])
phylum.T1               <- mutate(phylum.T1,v1_sum_percent = v1_sum/sum(v1_sum)) %>%
  arrange(desc(v1_sum_percent)) %>% slice(1:4) %>%
  mutate(phylum = factor(phylum,levels = levels(factor(phylum))[c(1,3,4,2)])) %>%                         
  select(c(1,starts_with("T1_"))) 

colnames(phylum.T1)     <- c("phylum",paste0("S",1:10))

phylum.long             <- melt(phylum.T1, id.vars="phylum",
                                measure.vars=grep("S\\d+", names(phylum.T1), val=T),
                                variable.name="sample",
                                value.name="value")

tiff("Figure1.tiff",width = 800,height = 800)
ggplot(data = phylum.long, aes(x = sample, y = value, fill = phylum)) + 
  geom_bar(stat='identity') + 
  facet_grid(~phylum) +
  theme(legend.position="none") +
  scale_fill_manual(values = c("coral1","gray","palegreen1","steelblue1")) +
  scale_y_continuous(limits = c(0,650)) +
  theme(strip.text.x = element_text(size = 15, colour = "black", angle = 00)) +
  labs(x = "Sample (Time point 1)") + labs(y = "Abundance") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 10)) +
  theme(axis.text.y = element_text(size = 10)) +
  theme(axis.text=element_text(size=15),
        axis.title=element_text(size=15))
dev.off()

rm(list=ls())

##############################################################
## Change of abundance of bacteria at phylum and genus rank ##
### Change of bacteria at phylum level
phylum                  <- read.csv("Megan_phylum.csv", header = T, stringsAsFactors = FALSE)
phylum$sum              <- rowSums(phylum[,-1])
phylum                  <- mutate(phylum,percent = sum/sum(sum)) %>%
  arrange(desc(percent)) %>% select(1:21)
rownames(phylum)        <- phylum$X
phylum                  <- select(phylum,-1)
colnames(phylum)        <- c(paste0("T1_",1:10),paste0("T2_",1:10))
phylum                  <- t(phylum)

### Reletive abundance
####  phylum                  <- phylum/rowSums(phylum)
Abundance               <- (phylum[11:20,] - phylum[1:10,]) %>% data.frame() %>% 
  select(1:4) %>% mutate(sample = rep(c("Healthy","Infected"),each = 5),
                         subject_id = paste0("S",1:10))

phylum.long             <- melt(Abundance,id.vars = c("subject_id","sample"),
                                variable.name = "phylum")

Order                   <- c("S2", "S1", "S4", "S5", "S9", "S3", "S6", "S7", "S10", "S8")
phylum.Act              <- subset(phylum.long,phylum == "Actinobacteria") %>%
  arrange(desc(value)) %>% 
  mutate(subject_id = factor(subject_id,levels = Order))

p1                      <- ggplot(phylum.Act, aes(x = subject_id, y = value, fill=sample)) +
  geom_bar(stat='identity') + facet_grid(~sample, scales="free") +
  theme(strip.text.x = element_text(size = 15,colour = "black", angle = 00)) +
  labs(x = "Actinobacteria",y = "Change of Abundance (Phylum)") + 
  ylim(-400, 260) + theme(axis.text=element_text(size=10),
                          axis.title=element_text(size=15),
                          legend.position='none',
                          axis.text.x = element_blank())

rm(list = ls()[!ls() %in% "p1"])

########################################
### Change of bacteria at genus level ##
genus                   <- read.csv("Megan_genus.csv", header = T, stringsAsFactors = FALSE)
genus$sum               <- rowSums(genus[,-1])
genus                   <- mutate(genus,percent = sum/sum(sum)) %>%
  arrange(desc(percent)) %>% select(1:21)
rownames(genus)         <- genus$X
genus                   <- select(genus,-1)
colnames(genus)         <- c(paste0("T1_",1:10),paste0("T2_",1:10))
genus                   <- t(genus)

### Reletive abundance
####  genus                   <- genus/rowSums(genus)
Abundance               <- (genus[11:20,] - genus[1:10,]) %>% data.frame() %>% 
  select(1:4) %>% mutate(sample = rep(c("Healthy","Infected"),each = 5),
                         subject_id = paste0("S",1:10))

genus.long              <- melt(Abundance,id.vars = c("subject_id","sample"),
                                variable.name = "genus")

Order                   <- c("S2", "S1", "S4", "S5", "S9", "S3", "S6", "S7", "S10", "S8")
genus.Act               <- subset(genus.long,genus == "Corynebacterium") %>%
  arrange(desc(value)) %>% 
  mutate(subject_id = factor(subject_id,levels = Order))

p2                      <- ggplot(genus.Act, aes(x = subject_id, y = value, fill=sample)) +
  geom_bar(stat='identity') + facet_grid(~sample, scales="free") +
  theme(strip.text.x = element_text(size = 15,colour = "black", angle = 00)) +
  labs(x = "Corynebacterium",y = "Change of Abundance (Genus)") + 
  ylim(-400, 260) + theme(axis.text=element_text(size=10),
                          axis.title=element_text(size=15),
                          legend.position='none',
                          axis.text.x = element_blank())

tiff("Figure3.tiff",width = 800,height = 800)
ggarrange(p1, p2,labels = c("A", "B"),ncol = 2, nrow = 1)
dev.off()

rm(list=ls())

################
## PCoA plots ##
### phylum
phylum                  <- read.csv("Megan_phylum.csv", header = TRUE, stringsAsFactors = FALSE)
rownames(phylum)        <- phylum$X
phylum                  <- select(phylum,-1) %>% t() %>% data.frame()
phylum                  <- data.frame(phylum, visit = rep(c(1,2,3),each = 10),
                                      group = rep(rep(c("Healthy","Infected"),each = 5),3))
rownames(phylum)        <- unlist(lapply(paste0("T",1:3,"_"), paste0,1:10))

phy.health.dist         <- vegdist(subset(phylum,visit != 3 & group =="Healthy",
                                          select = -c(visit,group)),method = "bray")

phy.infect.dist         <- vegdist(subset(phylum,visit != 3 & group =="Infected",
                                          select = -c(visit,group)),method = "bray")
Factor.level            <- gl(2,5)

beta.phy.health         <- betadisper(phy.health.dist, Factor.level, type = "median")
phy.health.loc          <- scores(beta.phy.health, display = "sites")
phy.health.PCoA1        <- phy.health.loc[, 1]
phy.health.PCoA2        <- phy.health.loc[, 2]

beta.phy.infect         <- betadisper(phy.infect.dist, Factor.level, type = "median")
phy.infect.loc          <- scores(beta.phy.infect, display = "sites")
phy.infect.PCoA1        <- phy.infect.loc[, 1]
phy.infect.PCoA2        <- phy.infect.loc[, 2]

### genus
genus                   <- read.csv("Megan_genus.csv", header = TRUE,
                                    stringsAsFactors = FALSE)
rownames(genus)         <- genus$X
genus                   <- select(genus,-1) %>% t() %>% data.frame()
genus                   <- data.frame(genus, visit = rep(c(1,2,3),each = 10),
                                      group = rep(rep(c("Healthy","Infected"),each = 5),3))
rownames(genus)         <- unlist(lapply(paste0("T",1:3,"_"), paste0,1:10))

gen.health.dist         <- vegdist(subset(genus,visit != 3 & group =="Healthy",
                                          select = -c(visit,group)),method = "bray")

gen.infect.dist         <- vegdist(subset(genus,visit != 3 & group =="Infected",
                                          select = -c(visit,group)),method = "bray")

beta.gen.health         <- betadisper(gen.health.dist, Factor.level, type = "median")
gen.health.loc          <- scores(beta.gen.health, display = "sites")
gen.health.PCoA1        <- gen.health.loc[, 1]
gen.health.PCoA2        <- gen.health.loc[, 2]

beta.gen.infect         <- betadisper(gen.infect.dist, Factor.level, type = "median")
gen.infect.loc          <- scores(beta.gen.infect, display = "sites")
gen.infect.PCoA1        <- gen.infect.loc[, 1]
gen.infect.PCoA2        <- gen.infect.loc[, 2]

###### Plot
# tiff("Figure4.tiff",width = 800,height = 900)
par(mfrow = c(2,2),mar = c(4,0,4,0),oma = c(4,4,2,2))

plot(beta.phy.health,ellipse = F, conf = 0.05, hull = F, cex = 0.4, col = c("blue","red"),
     segments = T, seg.lty = 5, label = F, label.cex = 0.5, xlim=c(-0.6,0.6), ylim=c(-0.5, 0.6),
     xlab ="PCoA 1", main = "",yaxt = 'n') 
text(phy.health.PCoA1, phy.health.PCoA2, rownames(phy.health.loc), cex = 0.8, col = c(rep("blue",5), rep("red",5)))
text(0,0.6,"Healthy",cex = 1)
axis(2,at = seq(-0.4,0.6,0.2))

plot(beta.phy.infect,ellipse = F, conf = 0.05, hull = F, cex = 0.4, col = c("blue","red"),
     segments = T, seg.lty = 5, label = F, label.cex = 0.5, xlim=c(-0.6,0.6), ylim=c(-0.5, 0.6),
     xlab ="PCoA 1", main = "",yaxt = 'n',ylab = "") 
text(phy.infect.PCoA1, phy.infect.PCoA2, rownames(phy.infect.loc), cex = 0.8, col = c(rep("blue",5), rep("red",5)))
text(0,0.6,"Infected",cex = 1)
text(0.4,0.55,"Time point 1: Blue")
text(0.4,0.50,"Time point 2: Red ")

plot(beta.gen.health,ellipse = F, conf = 0.05, hull = F, cex = 0.4, col = c("blue","red"),
     segments = T, seg.lty = 5, label = F, label.cex = 0.5, xlim=c(-0.6,0.6), ylim=c(-0.5, 0.6),
     xlab ="PCoA 1", main = "",yaxt = 'n') 
text(gen.health.PCoA1, gen.health.PCoA2, rownames(gen.health.loc), cex = 0.8, col = c(rep("blue",5), rep("red",5)))
text(0,0.6,"Healthy",cex = 1)
axis(2,at = seq(-0.4,0.6,0.2))

plot(beta.gen.infect,ellipse = F, conf = 0.05, hull = F, cex = 0.4, col = c("blue","red"),
     segments = T, seg.lty = 5, label = F, label.cex = 0.5, xlim=c(-0.6,0.6), ylim=c(-0.5, 0.6),
     xlab ="PCoA 1", main = "",yaxt = 'n',ylab = "") 
text(gen.infect.PCoA1, gen.infect.PCoA2, rownames(gen.infect.loc), cex = 0.8, col = c(rep("blue",5), rep("red",5)))
text(0,0.6,"Infected",cex = 1)

#### adjust the postion
mtext("PCoA 2", side = 2, outer = TRUE, cex = 0.8,las = 0,line = 2.5,padj = 0,adj = 0.76)
mtext("PCoA 2", side = 2, outer = TRUE, cex = 0.8,las = 0,line = 2.5,padj = 0,adj = 0.22)

#### padj to adjust the postion 
mtext("A",side = 2,outer = TRUE,cex = 1.5,las = 2,padj = -21,adj = 3.5)
mtext("B",side = 2,outer = TRUE,cex = 1.5,las = 2,padj =  2,adj = 3.5)
# dev.off()

## Permutation Test
phy.distance            <- dist(beta.phy.infect$centroids) - dist(beta.phy.health$centroids)
gen.distance            <- dist(beta.gen.infect$centroids) - dist(beta.gen.health$centroids)

Perm_N                  <- 100

phy.perm.data           <- subset(phylum,visit != 3,select = -c(visit,group))
gen.perm.data           <- subset(genus,visit != 3,select = -c(visit,group))

Permutation             <- function(Index,Data){
  set.seed(Index)
  Order                 <- sample(1:nrow(Data),nrow(Data))
  
  # Random Define the First Ten as health and the following as infected
  health.dist           <- vegdist(Data[Order[1:10],],method = "bray")
  
  infect.dist           <- vegdist(Data[Order[11:20],],method = "bray")
  
  beta.health           <- betadisper(health.dist, gl(2,5), type = "median")
  beta.infect           <- betadisper(infect.dist, gl(2,5), type = "median")
  
  return(dist(beta.infect$centroids) - dist(beta.health$centroids))
}

phy.perm.dist           <- sapply(1:Perm_N, Permutation,Data = phy.perm.data)
gen.perm.dist           <- sapply(1:Perm_N, Permutation,Data = gen.perm.data)

phy.pvalue              <- sum(phy.perm.dist >= phy.distance)/Perm_N
gen.pvalue              <- sum(gen.perm.dist >= gen.distance)/Perm_N

#### 
par(mfrow = c(1,1))

hist(phy.perm.dist, breaks = 20,xlab = "", 
     main = sprintf("Histogram of distance with %d permutations (Phylum)", Perm_N))

abline(v = quantile(phy.perm.dist, 0.95), col="blue", lwd=3)
text(quantile(phy.perm.dist, 0.95)-0.09, 12,"Cutoff = 0.05", col = "blue", lwd=3)

abline(v = phy.distance, col = "red", lwd = 2, lty = 3)
text(phy.distance + 0.11, 12,labels = sprintf("P-value = %.3f",phy.pvalue),
     col = "red", lwd=3)

hist(gen.perm.dist, breaks = 20,xlab = "", 
     main = sprintf("Histogram of distance with %d permutations (Genus)", Perm_N))

abline(v = quantile(gen.perm.dist, 0.95), col="blue", lwd=3)
text(quantile(gen.perm.dist, 0.95)-0.09, 12,"Cutoff = 0.05", col = "blue", lwd=3)

abline(v = gen.distance, col = "red", lwd = 2, lty = 3)
text(gen.distance, 12,labels = sprintf("P-value = %.3f",gen.pvalue),
     col = "red", lwd=3)
